### PR TITLE
Integration test app

### PR DESCRIPTION
**Pull request by jwworth**
This is a project to add better integration test coverage to Rails, per:
https://github.com/rails/rails/issues/20945

The idea is a Rails app inside Rails, with a test suite covering core Rails features, run as part of continuous integration.

TODO:
- [x] Create application
- [x] Add simple passing test
- [x] Point to local Rails in Gemfile
- [x] Integrate with CI

### CODE REVIEW

**Comment by matthewd** [comment]
What is this [envisaged to be] testing? :confused:

**Comment by jwworth** [comment]
Hey @sgrif -- I made a first pass at the test app we talked about. Please let me know what you think, and also if you know what I need to change for the test to run on Travis.

**Comment by sgrif** [comment]
@matthewd The goal is to add better integration test coverage. I think the best way to go about doing that is to have a Rails app, and have its test suite be part of our general test suite. I had asked that the first pass at it be just wiring it up, so we can add tests for each feature individually.

I don't envision this being fully comprehensive, but I eventually would like to see minimal happy path usage of each "feature" (a feature being roughly defined as each module mixed into the various base classes).

This was in response to the AC::Parameters change not breaking our build.

File: ci/travis.rb [hunk 1]
```diff
@@ -23,7 +23,8 @@ class Build
     'ar'       => 'activerecord',
     'av'       => 'actionview',
     'aj'       => 'activejob',
-    'guides'   => 'guides'
+    'guides'   => 'guides',
+    'ita'      => 'integration-test-app'
```
Comment (sgrif) [inline ci/travis.rb]: Can we add a trailing comma here so there's not churn on this line in the future?

File: integration-test-app/README.rdoc [hunk 2]
```diff
@@ -0,0 +1,28 @@
+== README
+
+This README would normally document whatever steps are necessary to get the
```
Comment (sgrif) [inline integration-test-app/README.rdoc]: Let's fill this in with what it actually is.

File: ci/travis.rb (hunk 1 above)
Comment (jwworth) [inline ci/travis.rb]: Yes, I'll make that change.

**Comment by matthewd** [comment]
I'm not thrilled at the idea of committing a copy of all the files that come out of the generator... if we do a good enough job of keeping them up to date, it'll be a pain to maintain.

But we also already have framework-wide integration tests -- which address my above concern by generating the app on demand -- e.g. https://github.com/rails/rails/blob/master/railties/test/application/per_request_digest_cache_test.rb

Should we not be extending that coverage, instead of inventing a parallel mechanism?

**Comment by sgrif** [comment]
My main concern with extending the railties tests to be more comprehensive is speed. I think the cost of maintaining a minimal rails app is much lower in this instance. I also think that the most natural way to integration test things is through a Rails app.

**Comment by matthewd** [comment]
Generating fit-to-purpose apps simplifies covering minor variations on a theme -- particularly once multi-level permutations get involved. A monolithic hand-crafted app, that sets out to be more "fully comprehensive" than AWDwR, sounds like quite the undertaking.

**Comment by sgrif** [comment]
I think that individual controllers can very much fill the same roll as fit-to-purpose apps for the purposes of testing. I agree this is a non-trivial undertaking, but I do think it is necessary. While AWDwR provides a lot of value, I feel strongly that we need an in-tree integration test suite, with the intention of being more broad (the AC::Parameters change did not fail the AWDwR suite for example).

**Comment by sgrif** [comment]
@matthewd Do you have any other thoughts? I'd like to merge this if there's no further objections. I still feel quite strongly that we need to improve our integration test coverage, and I think this is the most natural path to do that as per my previous comments.

**Comment by rafaelfranca** [comment]
Maybe before we decide if it make sense or not we need integrations tests on this application so we can see if it could be done in railties tests or if an application is needed.

**Comment by tenderlove** [comment]
I'm not super thrilled about checking in a fully blown app as well. Generating new apps is trivially parallelizable, not to mention we could generate an app once for many "integration" tests which would amortize the cost.  IOW I think there are other techniques for speed that have lower maintenance burden that we should explore.

I'd like to hear other folks opinions before this is merged.

**Comment by rafaelfranca** [comment]
> not to mention we could generate an app once for many "integration" tests which would amortize the cost

I think this is exactly what we are doing in the railties test. https://github.com/rails/rails/blob/ec94f00ba3cf250eb54fc5b7a5e3ed4b90164f34/railties/test/isolation/abstract_unit.rb#L108

**Comment by tenderlove** [comment]
@rafaelfranca oops! Well I suppose we should make sure it's actually working. ;-) since we actually do cache the generated app, is checking in a new app actually going to help? I'd really like to see information about real-world impact, especially when it comes to speed.

**Comment by pixeltrix** [comment]
I think having a single app trying to test all the features will quickly become unmaintainable - the routing tests for Action Dispatch used to be in a single routeset and that became huge so I dread to think what would happen with a whole app.

IIRC the current railties application tests run sequentially - maybe the effort would be better spent on making those parallel?

**Comment by rafaelfranca** [comment]
> IIRC the current railties application tests run sequentially - maybe the effort would be better spent on making those parallel?

You can also run them in parallel right now. `CORES=4 rake test` will run the tests in parallel. I'm closing since we are not going to merge this new application.

Thank you for working on this.

**Pull request closed without merging**